airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java

```diff
@@ -119,26 +119,21 @@ public ReplicationOutput run(final StandardSyncInput syncInput, final Path jobRo
         destination.start(destinationConfig, jobRoot);
         source.start(sourceConfig, jobRoot);
 
-        final Future<?> destinationOutputThreadFuture = executors.submit(getDestinationOutputRunnable(
-            destination,
-            cancelled,
-            destinationMessageTracker,
-            mdc));
-
-        final Future<?> replicationThreadFuture = executors.submit(getReplicationRunnable(
-            source,
-            destination,
-            cancelled,
-            mapper,
-            sourceMessageTracker,
-            mdc));
-
-        LOGGER.info("Waiting for source thread to join.");
-        replicationThreadFuture.get();
-        LOGGER.info("Source thread complete.");
-        LOGGER.info("Waiting for destination thread to join.");
-        destinationOutputThreadFuture.get();
-        LOGGER.info("Destination thread complete.");
+        final CompletableFuture<?> destinationOutputThreadFuture = CompletableFuture.runAsync(
+            getDestinationOutputRunnable(destination, cancelled, destinationMessageTracker, mdc),
+            executors);
+
+        final CompletableFuture<?> replicationThreadFuture = CompletableFuture.runAsync(
+            getReplicationRunnable(source, destination, cancelled, mapper, sourceMessageTracker, mdc),
+            executors);
+
+        LOGGER.info("Waiting for source and destination threads to complete!!");
+        // CompletableFuture#allOf waits until all futures finish before returning, even if one throws an
+        // exception. So in order to handle exceptions from a future immediately without needing to wait for
+        // the other future to finish, we first call CompletableFuture#anyOf.
+        CompletableFuture.anyOf(replicationThreadFuture, destinationOutputThreadFuture).get();
```

Comment:
Can we log between these so it's clear when it's triggered?

Reply:
Done